crates/devolutions-agent-shared/src/windows/registry.rs

```diff
@@ -0,0 +1,86 @@
+pub enum RegistryError {
+    #[error("failed to open registry key `{key}`")]
+    OpenKey { key: String, source: std::io::Error },
+    #[error("failed to enumerate registry key values from `{key}`")]
+    EnumKeyValues { key: String, source: std::io::Error },
+    #[error("failed to read registry value `{value}` from key `{key}`")]
+    ReadValue {
+        value: String,
+        key: String,
+        source: std::io::Error,
+    },
+    #[error(transparent)]
+    Uuid(#[from] UuidError),
+}
+
+/// Get the product code of an installed MSI using its upgrade code.
+pub fn get_product_code(update_code: &str) -> Result<Option<String>, RegistryError> {
+    let reversed_hex_uuid = uuid_to_reversed_hex(update_code)?;
+
+    let key_path = format!("{REG_CURRENT_VERSION}\\Installer\\UpgradeCodes\\{reversed_hex_uuid}");
+
+    let update_code_key = winreg::RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE).open_subkey(&key_path);
+
+    // Product not installed if no key found.
+    let update_code_key = match update_code_key {
+        Ok(key) => key,
+        Err(_) => return Ok(None),
+    };
+
+    // Product code is the name of the only value in the registry key.
+    let (product_code, _) = match update_code_key.enum_values().next() {
+        Some(value) => value.map_err(|source| RegistryError::EnumKeyValues { key: key_path, source })?,
+        None => return Ok(None),
+    };
+
+    Ok(Some(reversed_hex_to_uuid(&product_code)?))
+}
+
+/// Get the installed version of a product using Windows registry. Returns `None` if the product
+/// is not installed.
+pub fn get_installed_product_version(update_code: &str) -> Result<Option<DateVersion>, RegistryError> {
+    let product_code_uuid = match get_product_code(update_code)? {
+        Some(uuid) => uuid,
+        None => return Ok(None),
+    };
+
+    let key_path = format!("{REG_CURRENT_VERSION}\\Uninstall\\{product_code_uuid}");
+
+    const VERSION_VALUE_NAME: &str = "Version";
+
+    // Now we know the product code of installed MSI, we could read its version.
+    let product_tree = winreg::RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE)
+        .open_subkey(&key_path)
+        .map_err(|source| RegistryError::OpenKey {
+            key: key_path.clone(),
+            source,
+        })?;
+
+    let product_version: u32 =
+        product_tree
+            .get_value(VERSION_VALUE_NAME)
+            .map_err(|source| RegistryError::ReadValue {
+                value: VERSION_VALUE_NAME.to_owned(),
+                key: key_path,
+                source,
+            })?;
+
+    // Convert encoded MSI version number to human-readable date.
+    let short_year = (product_version >> 24) + 2000;
+    let month = (product_version >> 16) & 0xFF;
+    let day = product_version & 0xFFFF;
+
+    Ok(Some(DateVersion {
+        year: short_year,
+        month,
+        day,
+        // NOTE: Windows apps could only have 3 version numbers (major, minor, patch).
+        revision: 0,
```

Comment:
note: Hopefully should not cause any problem because we don’t use this number ourselves, but if we end up using the revision number one day, we’ll be basically reporting an older version regardless of what was actually installed… :thinking:
We may want to use a different approach at some point.